README.md
# gamegos-supervisor

[![Build Status](https://travis-ci.org/gamegos/supervisor-cookbook.svg?branch=master)](https://travis-ci.org/gamegos/supervisor-cookbook) [![Cookbook Version](https://img.shields.io/cookbook/v/gamegos-supervisor.svg)](https://supermarket.chef.io/cookbooks/gamegos-supervisor)

This cookbook installs and configures [Supervisor](https://github.com/Supervisor/supervisor).

# Requirements

## Platforms

- Ubuntu 14.04+
- Centos 7+
- Debian 7+

## Chef

- Chef 12+

## Cookbooks

- poise-python
- poise-service

# Usage

Here's a quick example of installing the supervisord and adding some programs.

```ruby
supervisor_install '3.3'

supervisor_config 'supervisor'

supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end

supervisor_process 'vi' do
  type 'program'
  command '/usr/bin/vi'
end

supervisor_group 'cat' do
  programs ['cat']
end

supervisor_service 'supervisor'
```

# Recipes

- **default** - installs the supervisord.

# Resources

### supervisor_install

Install Supervisor with pip.

#### Actions

- `:create`

#### Examples

```ruby
supervisor_install '3.3'
```

### supervisor_config

Define configurations of supervisord.

#### Actions

- `:create`

#### Properties

| Property                       |       Type        |                  Default Value                   |
| ------------------------------ | :---------------: | :----------------------------------------------: |
| `socket_file`                  |     `String`      |           `'/var/run/supervisor.sock'`           |
| `unix_http_server_chmod`       |     `String`      |                     `'700'`                      |
| `unix_http_server_chown`       |     `Integer`     |                  `'root:root'`                   |
| `unix_http_server_username`    |  `[String, Nil]`  |                      `nil`                       |
| `unix_http_server_password`    |  `[String, Nil]`  |                      `nil`                       |
| `supervisord_config_directory` |     `String`      |               `'/etc/supervisor'`                |
| `supervisord_log_directory`    |     `String`      |             `'/etc/log/supervisor'`              |
| `supervisord_logfile`          |     `String`      | `"#{supervisord_log_directory}/supervisord.log"` |
| `supervisord_logfile_maxbytes` |     `String`      |                     `'50mb'`                     |
| `supervisord_logfile_backups`  |     `Integer`     |                       `10`                       |
| `supervisord_loglevel`         |     `String`      |                     `'info'`                     |
| `supervisord_pidfile`          |     `String`      |           `'/var/run/supervisord.pid'`           |
| `supervisord_nodaemon`         |     `Boolean`     |                     `false`                      |
| `supervisord_minfds`           |     `Integer`     |                      `1024`                      |
| `supervisord_minprocs`         |     `Integer`     |                      `200`                       |
| `supervisord_nocleanup`        |     `Boolean`     |                     `false`                      |
| `supervisord_user`             |  `[String, Nil]`  |                      `nil`                       |
| `supervisord_umask`            |  `[String, Nil]`  |                      `nil`                       |
| `supervisord_identifier`       |     `String`      |                  `'supervisor'`                  |
| `supervisord_strip_ansi`       |     `Boolean`     |                     `false`                      |
| `supervisord_environment`      |      `Hash`       |                       `{}`                       |
| `inet_port`                    |     `String`      |                 `'0.0.0.0:9001'`                 |
| `inet_username`                |  `[String, Nil]`  |                      `nil`                       |
| `inet_password`                |  `[String, Nil]`  |                      `nil`                       |
| `include_files`                | `[String, Array]` |    `"#{supervisord_config_directory}/*.conf"`    |
| `action`                       |     `Symbol`      |                    `:create`                     |

#### Examples

```ruby
# Create a custom config
supervisor_config 'supervisor' do
  supervisord_config_directory '/etc/supervisor'
  socket_file '/run/supervisor.sock'
  inet_port '0.0.0.0:9010'
  inet_username 'randy'
  inet_password 'elite'
  supervisord_environment NODE_CONFIG_PORT: '5000', NODE_CONFIG_KEY: '1DFS123SDFK'
  include_files [ '/etc/supervisor/*.conf', '/etc/supervisor/*.ini' ]
  action :create
end
```

```ruby
# Create a default config
supervisor_config 'supervisor'
```

#### Notes

The name property of the resource should always just be `supervisor`.

### supervisor_process

Creates a process for supervisor. The process may be a 'program', 'eventlistener' or 'fcgi-program'.

#### Actions

- `:create`
- `:delete`

#### Properties

| Property                       |        Type        |                    Default Value                     |
| ------------------------------ | :----------------: | :--------------------------------------------------: |
| `type` (@required)             |      `String`      |                          ``                          |
| `command` (@required)          |      `String`      |                          ``                          |
| `process_name`                 |      `String`      |                 `'%(program_name)s'`                 |
| `numprocs`                     |     `Integer`      |                         `1`                          |
| `numprocs_start`               |     `Integer`      |                         `0`                          |
| `priority`                     |     `Integer`      |                        `999`                         |
| `autostart`                    |     `Boolean`      |                       `false`                        |
| `autorestart`                  |     `Boolean`      |                       `false`                        |
| `startsecs`                    |     `Integer`      |                         `1`                          |
| `startretries`                 |     `Integer`      |                         `3`                          |
| `exitcodes`                    |      `Array`       |                       `[0, 2]`                       |
| `stopsignal`                   | `[String, Symbol]` |                       `:TERM`                        |
| `stopwaitsec`                  |     `Integer`      |                         `10`                         |
| `stopasgroup`                  |     `Boolean`      |                        `true`                        |
| `killasgroup`                  |     `Boolean`      |                        `true`                        |
| `user`                         |      `String`      |                        `nil`                         |
| `redirect_stderr`              |     `Boolean`      |                       `false`                        |
| `environment`                  |       `Hash`       |                         `{}`                         |
| `directory`                    |  `[String, Nil]`   |                        `nil`                         |
| `umask`                        |  `[String, Nil]`   |                        `nil`                         |
| `serverurl`                    |      `String`      |                       `'AUTO'`                       |
| `stdout_logfile`               |      `String`      |                       `'AUTO'`                       |
| `stdout_logfile_maxbytes`      |      `String`      |                       `'50mb'`                       |
| `stdout_logfile_backups`       |     `Integer`      |                         `10`                         |
| `stdout_capture_maxbytes`      |      `String`      |                        `'0'`                         |
| `stdout_events_enabled`        |     `Boolean`      |                       `false`                        |
| `stderr_logfile`               |      `String`      |                       `'AUTO'`                       |
| `stderr_logfile_maxbytes`      |      `String`      |                       `'50mb'`                       |
| `stderr_logfile_backups`       |     `Integer`      |                         `10`                         |
| `stderr_capture_maxbytes`      |      `String`      |                        `'0'`                         |
| `stderr_events_enabled`        |     `Boolean`      |                       `false`                        |
| `eventlistener_buffer_size`    |     `Integer`      |                         `10`                         |
| `eventlistener_events`         |      `Array`       |                     `['EVENT']`                      |
| `eventlistener_result_handler` |      `String`      |      `'supervisor.dispatchers:default_handler'`      |
| `fcgi_socket`                  |      `String`      | `'unix:///var/run/supervisor/%(program_name)s.sock'` |
| `fcgi_socket_owner`            |  `[String, Nil]`   |                        `nil`                         |
| `fcgi_socket_mode`             |      `String`      |                        `0700`                        |
| `action`                       |      `Symbol`      |                      `:create`                       |

#### Examples

```ruby
supervisor_process 'my-nginx' do
  type 'program'
  command "/usr/sbin/nginx -g 'daemon off;'"
  numprocs 5
  process_name "%(program_name)s_%(process_num)02d"
  autorestart true
  user 'root'
end
```

```ruby
supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end
```

### supervisor_group

Creates a group for supervisor.

#### Actions

- `:create`
- `:delete`

#### Properties

| Property               |   Type    | Default Value |
| ---------------------- | :-------: | :-----------: |
| `programs` (@required) |  `Array`  |      ``       |
| `priority`             | `Integer` |     `999`     |
| `action`               | `Symbol`  |   `:create`   |


#### Examples

```ruby
supervisor_group 'my-web-servers' do
  programs [ 'my-nginx', 'my-apache' ]
  priorty 100
end
```

### supervisor_service

Creates Supervisor service for your systems.

#### Actions

- `:create`
- `:reload`

#### Properties


| Property                      |   Type   |                  Default Value                   |
| ----------------------------- | :------: | :----------------------------------------------: |
| `supervisord_executable_path` | `String` | `node['supervisor']['supervisord_default_path']` |
| `action`                      | `Symbol` |                    `:create`                     |

#### Examples

```ruby
supervisor_service 'supervisor'
```

#### Note

The name property of the resource should always just be `supervisor`.


CHANGELOG.md
# gamegos-supervisor CHANGELOG

This file is used to list changes made in each version of the gamegos-supervisor cookbook.

# 1.0.0

Update

- Add Chef 12 support
- Add tests
  - Testing Ubuntu and CentOS
- Update documentation
- Fixes some bugs


# 0.1.0

Initial release.

- Support Chef13+



recipes/test.rb
#
# Cookbook:: gamegos-supervisor
# Recipe:: test
#
# Copyright:: 2018, The Authors, All Rights Reserved.
supervisor_install '3.3'

supervisor_config 'supervisor'

supervisor_process 'cat' do
  type 'program'
  command '/bin/cat'
end

supervisor_process 'vi' do
  type 'program'
  command '/usr/bin/vi'
end

supervisor_group 'cat' do
  programs ['cat']
end

supervisor_group 'vi' do
  programs ['vi']
end

supervisor_service 'supervisor'


test/vagrantfile_merge.rb
# This file is merged with the Vagrant configuration generated by test kitchen
Vagrant.configure('2') do |config|
  config.vbguest.auto_update = false if Vagrant.has_plugin?('vagrant-vbguest')
end


test/cookbooks/test/metadata.rb
name              'test'
maintainer        'Fatih Sarhan'
maintainer_email  'f9n@protonmail.com'
license           'GPL-3.0'
description       'Testing cookbook for gamegos-supervisor'
version           '0.0.1'
depends           'gamegos-supervisor'


metadata.rb
name 'gamegos-supervisor'
maintainer 'Fatih Sarhan'
maintainer_email 'f9n@protonmail.com'
license 'GPL-3.0'
description 'Installs/Configures supervisor'
long_description 'Installs/Configures supervisor'
chef_version '>= 12.0'
%w(amazon centos debian redhat ubuntu).each do |os|
  supports os
end
version '1.0.1'

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
issues_url 'https://github.com/gamegos/supervisor-cookbook/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
source_url 'https://github.com/gamegos/supervisor-cookbook'

depends 'poise-python', '~> 1.7.0'
depends 'poise-service', '~> 1.5.2'


resources/group.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_group

property :group_name, String, name_property: true
property :programs, Array, required: true
property :priority, Integer, default: 999
property :template, String, default: 'gamegos-supervisor'

action :create do
  declare_resource(:template, "supervisor_group_#{new_resource.group_name}") do
    cookbook new_resource.template
    path(lazy { "#{node.run_state['supervisor']['directory']}/group_#{new_resource.group_name}.conf" })
    source 'group.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables group: new_resource
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end

action :delete do
  file 'delete specific group configuration file' do
    path(lazy { "#{node.run_state['supervisor']['directory']}/group_#{new_resource.group_name}.conf" })
    action :delete
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end


resources/process.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_process

property :defined_process_name, String, name_property: true
property :type, String, required: true, equal_to: ['program', 'eventlistener', 'fcgi-program']
# Program
property :command, String, required: true
property :process_name, String, default: '%(program_name)s'
property :numprocs, Integer, default: 1
property :numprocs_start, Integer, default: 0
property :priority, Integer, default: 999
property :autostart, [TrueClass, FalseClass], default: false
property :autorestart, [TrueClass, FalseClass], default: false
property :startsecs, Integer, default: 1
property :startretries, Integer, default: 3
property :exitcodes, Array, default: [0, 2]
property :stopsignal, [String, Symbol], default: :TERM
property :stopwaitsecs, Integer, default: 10
property :stopasgroup,  [TrueClass, FalseClass], default: true
property :killasgroup,  [TrueClass, FalseClass], default: true
property :user, [String, NilClass], default: nil
property :redirect_stderr, [TrueClass, FalseClass], default: false
property :environment, Hash, default: {}
property :directory, [String, NilClass], default: nil
property :umask, [String, NilClass], default: nil
property :serverurl, String, default: 'AUTO'
property :stdout_logfile, String, default: 'AUTO'
property :stdout_logfile_maxbytes, String, default: '50MB'
property :stdout_logfile_backups, Integer, default: 10
property :stdout_capture_maxbytes, String, default: '0'
property :stdout_events_enabled,  [TrueClass, FalseClass], default: false
property :stderr_logfile, String, default: 'AUTO'
property :stderr_logfile_maxbytes, String, default: '50MB'
property :stderr_logfile_backups, Integer, default: 10
property :stderr_capture_maxbytes, String, default: '0'
property :stderr_events_enabled, [TrueClass, FalseClass], default: false

# Eventlistener
property :eventlistener_buffer_size, Integer, default: 10
property :eventlistener_events, Array, default: ['EVENT']
property :eventlistener_result_handler, String, default: 'supervisor.dispatchers:default_handler'

# Fcgi Program
property :fcgi_socket, String, default: 'unix:///var/run/supervisor/%(program_name)s.sock'
property :fcgi_socket_owner, [String, NilClass], default: nil
property :fcgi_socket_mode, String, default: '0700'

property :template, String, default: 'gamegos-supervisor'

action :create do
  clean_name = new_resource.defined_process_name.downcase.tr(' ', '_')
  unique_name_for_process = "#{new_resource.type}_#{clean_name}"
  declare_resource(:template, "supervisor_#{unique_name_for_process}") do
    cookbook new_resource.template
    path(lazy { "#{node.run_state['supervisor']['directory']}/#{unique_name_for_process}.conf" })
    source 'process.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables(
      name: clean_name,
      service: new_resource
    )
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end

action :delete do
  clean_name = new_resource.defined_process_name.downcase.tr(' ', '_')
  unique_name_for_process = "#{new_resource.type}_#{clean_name}"
  file 'delete specific program configuration file' do
    path(lazy { "#{node.run_state['supervisor']['directory']}/#{unique_name_for_process}.conf" })
    action :delete
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end


resources/config.rb
# To learn more about Custom Resources, see https://docs.chef.io/custom_resources.html
resource_name :supervisor_config

property :supervisord_config_directory, String, default: '/etc/supervisor'
property :socket_file, String, default: '/var/run/supervisor.sock'

property :unix_http_server_chmod, String, default: '700'
property :unix_http_server_chown, String, default: 'root:root'
property :unix_http_server_username, [String, NilClass], default: nil
property :unix_http_server_password, [String, NilClass], default: nil

property :supervisord_log_directory, String, default: '/var/log/supervisor'
property :supervisord_logfile, String, default: lazy {
  "#{supervisord_log_directory}/supervisord.log"
}
property :supervisord_logfile_maxbytes, String, default: '50MB'
property :supervisord_logfile_backups, Integer, default: 10
property :supervisord_loglevel, String, default: 'info'
property :supervisord_pidfile, String, default: '/var/run/supervisord.pid'
property :supervisord_nodaemon, [TrueClass, FalseClass], default: false
property :supervisord_minfds, Integer, default: 1024
property :supervisord_minprocs, Integer, default: 200
property :supervisord_nocleanup, [TrueClass, FalseClass], default: false
property :supervisord_user, [String, NilClass], default: nil
property :supervisord_umask, [String, NilClass], default: nil
property :supervisord_identifier, String, default: 'supervisor'
property :supervisord_strip_ansi, [TrueClass, FalseClass], default: false
property :supervisord_environment, Hash, default: {}

property :inet_port, [String, NilClass], default: '0.0.0.0:9001'
property :inet_username, [String, NilClass], default: nil
property :inet_password, [String, NilClass], default: nil

property :include_files, [String, Array], default: lazy {
  "#{supervisord_config_directory}/*.conf"
}

property :template, String, default: 'gamegos-supervisor'

action :create do
  supervisor_config_directory = new_resource.supervisord_config_directory
  supervisor_config_file = "#{supervisor_config_directory}/supervisord.conf"
  with_run_context :root do
    node.run_state['supervisor'] ||= {}
    node.run_state['supervisor']['directory'] = supervisor_config_directory
    node.run_state['supervisor']['config_file'] = supervisor_config_file
  end
  node.default['supervisor']['inet_http_server']['inet_username'] = new_resource.inet_username
  node.default['supervisor']['inet_http_server']['inet_password'] = new_resource.inet_password
  node.default['supervisor']['inet_http_server']['inet_port'] = new_resource.inet_port

  directory supervisor_config_directory do
    owner 'root'
    group 'root'
    mode '755'
    recursive true
  end

  directory new_resource.supervisord_log_directory do
    owner 'root'
    group 'root'
    mode '755'
    recursive true
  end

  declare_resource(:template, 'supervisord_config_file') do
    cookbook new_resource.template
    path supervisor_config_file
    source 'supervisord.conf.erb'
    owner 'root'
    group 'root'
    mode '644'
    variables config: new_resource
    notifies :reload, find_resource(:supervisor_service, 'supervisor'), :delayed
  end
end
